File: n4/5_4.cpp
#include <iostream>
#include <map>
#include <set>
#include <string>

using namespace std;

int main() {
    int n;
    cin >> n;

    map<string, string> regions;
    set<string> deleted_names;

    for (int i = 0; i < n; ++i) {
        string command;
        cin >> command;

        if (command == "CHANGE") {
            string region, new_center;
            cin >> region >> new_center;

            if (regions.count(region)) {
                string old_center = regions[region];
                if (old_center == new_center) {
                    cout << "Incorrect" << endl; //новый = старый
                } else {
                    regions[region] = new_center;
                    cout << "Region " << region << " has changed its administrative center from "
                         << old_center << " to " << new_center << endl;
                }
            }
            else if (deleted_names.count(region)) {
                cout << "Incorrect" << endl; //переименнованый нельзя использ 
            }
            else {
                regions[region] = new_center;
                cout << "New region " << region << " with administrative center " << new_center << endl;
            }

        } else if (command == "RENAME") {
            string old_region, new_region;
            cin >> old_region >> new_region;

            if (old_region == new_region || !regions.count(old_region) || regions.count(new_region)) {
                cout << "Incorrect" << endl; //если = старый или не сущ или новый регион использ
            } else {
                regions[new_region] = regions[old_region];
                regions.erase(old_region);
                deleted_names.insert(old_region);
                cout << old_region << " has been renamed to " << new_region << endl;
            }

        } else if (command == "ABOUT") {
            string region;
            cin >> region;

            if (regions.count(region)) {
                cout << region << " has administrative center " << regions[region] << endl;
            } else {
                cout << "Incorrect" << endl; //если регион не найден
            }

        } else if (command == "ALL") {
            for (const auto& item : regions) {
                cout << item.first << " - " << item.second << endl;
            }
        }
    }

    return 0;
}

File: n4/5_4.rs
use std::collections::{HashMap, HashSet};
use std::io::{self, Write};

fn main() {
    let mut input = String::new();
    let mut regions: HashMap<String, String> = HashMap::new();
    let mut deleted_names: HashSet<String> = HashSet::new();

    io::stdin().read_line(&mut input).unwrap();
    let n: usize = input.trim().parse().unwrap();
    input.clear();

    for _ in 0..n {
        print!("");
        io::stdout().flush().unwrap();
        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        let mut parts = input.trim().split_whitespace();
        let command = parts.next().unwrap();

        match command {
            "CHANGE" => {
                let region = parts.next().unwrap().to_string();
                let new_center = parts.next().unwrap().to_string();

                if let Some(old_center_ref) = regions.get(&region) {
                    let old_center = old_center_ref.clone(); // клонируем, чтобы завершить immutable borrow
                    if old_center == new_center {
                        println!("Incorrect");
                    } else {
                        regions.insert(region.clone(), new_center.clone());
                        println!(
                            "Region {} has changed its administrative center from {} to {}",
                            region, old_center, new_center
                        );
                    }
                } else if deleted_names.contains(&region) {
                    println!("Incorrect");
                } else {
                    regions.insert(region.clone(), new_center.clone());
                    println!(
                        "New region {} with administrative center {}",
                        region, new_center
                    );
                }
            }

            "RENAME" => {
                let old_region = parts.next().unwrap().to_string();
                let new_region = parts.next().unwrap().to_string();

                if old_region == new_region
                    || !regions.contains_key(&old_region)
                    || regions.contains_key(&new_region)
                {
                    println!("Incorrect");
                } else {
                    let center = regions.remove(&old_region).unwrap();
                    regions.insert(new_region.clone(), center);
                    deleted_names.insert(old_region.clone());
                    println!("{} has been renamed to {}", old_region, new_region);
                }
            }

            "ABOUT" => {
                let region = parts.next().unwrap().to_string();
                if let Some(center) = regions.get(&region) {
                    println!("{} has administrative center {}", region, center);
                } else {
                    println!("Incorrect");
                }
            }

            "ALL" => {
                for (region, center) in &regions {
                    println!("{} - {}", region, center);
                }
            }

            _ => println!("Unknown command"),
        }
    }
}
